js/app.js

```diff
@@ -81,20 +68,98 @@ TextApp.prototype.setTheme = function() {
 };
 
+/**
+ * Remove the editor so it can be reinitialized.
+ * @param editor The DOM element containing the editor.
+ */
+TextApp.prototype.removeEditor = function(editor) {
+  // Let the object do any clean up it needs.
+  if (this.editor_ !== null) {
+    this.editor_.destory();
+  }
+
+  // Clear the DOM
+  while (editor.firstElementChild !== null) {
+    editor.firstElementChild.remove();
+  }
+};
+
 /**
  * Called when all the services have started and settings are loaded.
  */
 TextApp.prototype.onSettingsReady_ = function() {
+  this.settingsController_ = new SettingsController(this.settings_);
+
+  this.initEditor_();
+
+  this.analytics_.setEnabled(this.settings_.get('analytics'));
+  this.analytics_.reportSettings(this.settings_);
+  this.windowController_.setAlwaysOnTop(this.settings_.get('alwaysontop'));
+
+  chrome.runtime.getBackgroundPage(function(bg) {
+    bg.background.onWindowReady(this);
+  }.bind(this));
+};
+
+/**
+ * Create a new editor and load all settings.
+ */
+TextApp.prototype.initEditor_ = function() {
+  // Remove any editor that already exists.
+  if (this.editor_ !== null) {
+    const editor = document.getElementById('editor');
+    this.removeEditor(editor);
+  }
+
+  if (this.settings_.get('screenreadermode')) {
+    this.editor_ = new EditorTextArea(editor, this.settings_);
+  } else {
+    this.editor_ = new EditorCodeMirror(editor, this.settings_);
+  }
+
+  if (!this.tabs_) {
+    // If tabs doesn't exist this is the first editor being created, if so
+    // create all the needed controllers.
+    this.dialogController_ = new DialogController(
+        $('#dialog-container'), this.editor_);
+    this.tabs_ = new Tabs(this.editor_, this.dialogController_, this.settings_);
+    this.menuController_ = new MenuController(this.tabs_);
+    this.windowController_ = new WindowController(
+        this.editor_, this.settings_, this.analytics_, this.tabs_);
+    this.hotkeysController_ = new HotkeysController(this.windowController_,
+        this.tabs_, this.editor_, this.settings_, this.analytics_);
+    this.searchController_ = new SearchController(this.editor_.getSearch());
+  } else {
+    // If tabs already exists, just replace the editor rather then creating a
+    // new Tabs object instance, this way we don't lose any tabs that were open.
+    this.tabs_.updateEditor(this.editor_);
+
+    // Controllers should be only created once.
+    // On any subsequent editor changes they should be notified of the editor
+    // change rather then reconstructed. This is to prevent these objects from
+    // creating spurious event handlers that all run in tandem.
+    this.windowController_.updateEditor(this.editor_);
+    this.hotkeysController_.updateEditor(this.editor_);
+    this.searchController_.updateEditor(this.editor_);
+  }
+
+  // Unlock all settings.
+  this.settings_.enableAll();
+
+  // Lock any settings the editor doesn't support.
+  const lockedSettings = this.editor_.lockedSettings();
```

Comment:
so if a user switches to screenreader mode then the settings will be changed to what is supported and disabled. then if they switch back their settings will be permanently altered? i think it would be preferable (particularly in the case where the user accidentally clicks this then switches it back off immediately) if the previous settings were restored. perhaps this could be a todo/follow up?

Reply:
Yeah i thought of this too, It's a bit finicky getting the settings state to restore correctly so i left it out of this initial PR. 

Will add a todo!